components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -3104,6 +3103,25 @@ private GroupsGetResponse listGroups(int startIndex, Integer count, String sortB
                 groupNames = getRoleNamesForGroupsEndpoint(domainName);
             }
 
+            totalGroupCount = groupNames.size();
+            // Adjust startIndex and endIndex to ensure they are within bounds
+            if (startIndex > 0) {
+                if (startIndex > totalGroupCount) {
+                    startIndex = totalGroupCount;
+                } else {
+                    startIndex -= 1;
+                }
+            } else {
+                startIndex = 0;
+            }
+            int endIndex;
+            if (count == null) {
+                endIndex = totalGroupCount;
+            } else {
+                endIndex = Math.min(startIndex + count, groupNames.size());
+            }
+            groupNames = new HashSet<>(new ArrayList<>(groupNames).subList(startIndex, endIndex));
```

Comment:
```suggestion
            groupNames = groupNames.stream()
                    .skip(startIndex)
                    .limit(endIndex - startIndex)
                    .collect(Collectors.toCollection(LinkedHashSet::new));
```

By this we can avoid redundant ArrayList creation

Reply:
Fixed by - https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/620/commits/7385e8b1c4ec54afa669f162e2e57e54a5776b25